compiler/damlc/daml-stdlib-src/DA/Exception.daml

```diff
@@ -0,0 +1,74 @@
+{-# LANGUAGE CPP #-}
+
+#ifndef DAML_EXCEPTIONS
+
+-- | HIDE
+module DA.Exception where
+
+#else
+
+-- | Exception handling.
+module DA.Exception
+    ( AnyException (..)
+    , Exception (..)
+    , displayException
+    , throw
+    , throwPure
+    , ActionCatch (..)
+    ) where
+
+import DA.Record
+import DA.Internal.LF
+import DA.Internal.Exception
+
+instance HasField "message" AnyException Text where
+    getField AnyException{..} = message
+    setField message AnyException{payload} = AnyException{..}
+
+instance HasField "payload" AnyException Any where
+    getField AnyException{..} = payload
+    setField payload AnyException{message} = AnyException{..}
+
+-- | Exception typeclass. This should not be implemented directly,
+-- instead, use the `exception` syntax.
+class Exception t where
+    toAnyException : t -> AnyException
+    fromAnyException : AnyException -> Optional t
+
+instance Exception AnyException where
+    toAnyException = identity
+    fromAnyException = Some
+
+-- | Get the error message associated with an exception.
+displayException : Exception e => e -> Text
+displayException e = getField @"message" (toAnyException e)
+
+-- | Throw an exception.
+throw : (Action m, Exception e) => e -> m t
+throw e = do
+    pure ()
```

Comment:
I assume this is intentional to work around the slighly odd behavior of `do` blocks? Perhaps worth adding a comment and maybe using `>>=` to make it even more explicit what we’re going for here (`>>` wouldn’t work if we hadn’t hacked around it in the compiler).